sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -95,7 +100,11 @@ class SparkOptimizer(
       EliminateLimits,
       ConstantFolding),
     Batch("User Provided Optimizers", fixedPoint, experimentalMethods.extraOptimizations: _*),
-    Batch("Replace CTE with Repartition", Once, ReplaceCTERefWithRepartition)))
+    Batch("Replace CTE with Repartition", Once, ReplaceCTERefWithRepartition),
+    Batch("Merge With expression", fixedPoint, MergeWithExpression),
+    Batch("Rewrite With expression", fixedPoint,
```

Comment:
how many places do we need to put this batch?

Reply:
Before Extract Python UDFs, infer filters and last of each optimizer.

Reply:
Can you elaborate on the issue if don't rewrite `With` before Extract Python UDFs and infer filters?

Reply:
Extract Python UDFs will extracts PythonUDFs from operators  to `BatchEvalPython`/`ArrowEvalPython`, so it is necessary to ensure that EvalPython does not contain common expression ref.

Reply:
There seems to be no problem with infer filters. As long as ensure infer filters once, I will fix it.

Reply:
OK now this makes me a bit worried. One major limitation of `With` is that we must move it as a whole. Otherwise we have dangling `CommonExpressionRef`.

It seems safer to rewrite `With` immediately at the beginning of the optimizer and right after the place that we produce new `With`.

Reply:
Yes, we must move `With` as a whole, so rewrite `With` every time after push down predicate?

Reply:
Yea, for now we are sure that only two places can produce `With`: the analyzer and the predicate pushsdown rule. It's safer to take care of these two.